include/cuco/detail/hyperloglog/finalizer.cuh

```diff
@@ -0,0 +1,163 @@
+
+  __host__ __device__ static double constexpr alpha_mm() noexcept
+  {
+    if constexpr (m == 16) {
+      return 0.673 * m * m;
+    } else if constexpr (m == 32) {
+      return 0.697 * m * m;
+    } else if constexpr (m == 64) {
+      return 0.709 * m * m;
+    } else {
+      return (0.7213 / (1.0 + 1.079 / m)) * m * m;
+    }
+  }
+
+  __host__ __device__ static double constexpr bias_corrected_estimate(double e) noexcept
+  {
+    if constexpr (Precision < 19) {
+      if (e < 5.0 * m) { return e - bias(e); }
+    }
+    return e;
+  }
+
+  __host__ __device__ static double constexpr bias(double e) noexcept
+  {
+    auto const anchor_index = interpolation_anchor_index(e);
+    int const n             = raw_estimate_data<Precision>().size();
+
+    auto low  = cuda::std::max(anchor_index - k + 1, 0);
+    auto high = cuda::std::min(low + k, n);
+    // Keep moving bounds as long as the (exclusive) high bound is closer to the estimate than
+    // the lower (inclusive) bound.
+    while (high < n and distance(e, high) < distance(e, low)) {
+      low += 1;
+      high += 1;
+    }
+
+    auto const& biases = bias_data<Precision>();
+    double bias_sum    = 0.0;
+    for (int i = low; i < high; ++i) {
+      bias_sum += biases[i];
+    }
+
+    return bias_sum / (high - low);
+  }
+
+  __host__ __device__ static double distance(double e, int i) noexcept
+  {
+    auto const diff = e - raw_estimate_data<Precision>()[i];
+    return diff * diff;
+  }
+
+  __host__ __device__ static int interpolation_anchor_index(double e) noexcept
+  {
+    auto const& estimates = raw_estimate_data<Precision>();
+    int left              = 0;
+    int right             = static_cast<int>(estimates.size()) - 1;
+    int mid;
+    int candidate_index = 0;  // Index of the closest element found
+
+    while (left <= right) {
+      mid = left + (right - left) / 2;
+
+      if (estimates[mid] < e) {
+        left = mid + 1;
+      } else if (estimates[mid] > e) {
+        right = mid - 1;
+      } else {
+        // Exact match found, no need to look further
+        return mid;
+      }
+    }
+
+    // At this point, 'left' is the insertion point. We need to compare the elements at 'left' and
+    // 'left - 1' to find the closest one, taking care of boundary conditions.
+
+    // Distance from 'e' to the element at 'left', if within bounds
+    double const dist_lhs = left < static_cast<int>(estimates.size())
+                              ? cuda::std::abs(estimates[left] - e)
+                              : cuda::std::numeric_limits<double>::max();
+    // Distance from 'e' to the element at 'left - 1', if within bounds
+    double const dist_rhs = left - 1 >= 0 ? cuda::std::abs(estimates[left - 1] - e)
+                                          : cuda::std::numeric_limits<double>::max();
+
+    candidate_index = (dist_lhs < dist_rhs) ? left : left - 1;
+
+    return candidate_index;
+  }
+};
```

Comment:
This is the crucial part where we have to match Spark's implementation exactly. I would appreciate if someone from the SPark team could review this section.

Reply:
> This is the crucial part where we have to match Spark's implementation exactly.

This doesn't seem like an appropriate requirement to fundamentally bake into the data structure. By all means, the design should be flexible enough to enable a user to provide sufficient customization to match whatever behavior they need, but we shouldn't be hardcoding matching Spark behavior into cuco data structures.

Reply:
I should clarify that the bias correction step in Spark is a 1:1 implementation of the ideas in the vanilla HLL++ paper. There aren't any Spark-specific alterations in the code. I was just following their implementation for better code comparability.

I would disagree in making this part a user-facing customization point for two reasons:
1. There aren't any optimization opportunities since it's single-threaded code and can be computed in no time.
2. HLL++ is already SOTA so there aren't any other options to replace this bias correction step with. If there is a better approach in the future, we can think of exposing this customization point. I already placed this step into a separate struct, so it's easy to replace.